### PR TITLE
[ENH]: fill_between extended to 3D

### PR DESCRIPTION
**Pull request by scottshambaugh**
## PR summary
Closes https://github.com/matplotlib/matplotlib/issues/28142

This extends the 2D `fill_between` method to 3D plots, and IMO works quite nicely. I’m finding it really powerful to generate surfaces easily. Inspired by @artmenlope's solution here: https://github.com/artmenlope/matplotlib-fill_between-in-3D/blob/master/FillBetween3d.py

Here's the what's new image and `fillbetween3d` gallery example: 
whatsnew

Note that I moved the `poly3d` example https://matplotlib.org/3.9.0/gallery/mplot3d/polys3d.html to a new example `fillunder3d`, to use this new method. Then I made a new example for `poly3d` since adding 3D collections wasn't shown anywhere else afaik.

New `fillunder3d` example:
fillbetween3d

New `poly3d` example:
whatsnewpoly

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by scottshambaugh** [comment]
Note that the white seams between the patches seems to be an instance of https://github.com/matplotlib/matplotlib/issues/1188, which I don't see a bug report for in 3D rendering. Is rather annoying but doesn't seem like there are easy fixes.

**Comment by scottshambaugh** [comment]
Note that I decided to not implement the "mode 2" idea in https://github.com/artmenlope/matplotlib-fill_between-in-3D/blob/master/FillBetween3d.py, which concatenates all the points and plots them as one polygon. I think the only benefit is getting rid of the white seams between the patches in the filled flat curves example, but it is degenerate in the general case, and this use case can be handled by users defining their own polygons such as in the new `poly3d` example.

For example, Mode 1 / Mode 2:
image

**Comment by scottshambaugh** [comment]
Here's the new plot types thumbnail:
image

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -1904,6 +1904,80 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
 
     plot3D = plot
 
+    def fill_between(self, xs1, ys1, zs1, xs2, ys2, zs2, *, where=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: I like without `s` better, and it matches the 2D fill_between. Switched

**Review by timhoffm — APPROVED** [review]
Only doc suggestions.

File: galleries/examples/mplot3d/polys3d.py [hunk 2]
```diff
@@ -1,47 +1,36 @@
...
-=============================================
+====================
+Generate 3D polygons
+====================
 
-Demonstrate how to create polygons which fill the space under a line
-graph. In this example polygons are semi-transparent, creating a sort
-of 'jagged stained glass' effect.
+Demonstrate how to create polygons in 3D. Here we stack 3 hexagons, somewhat
+like a traffic light.
```
[the remark this answers is not in the document]
Comment (scottshambaugh) [reply galleries/examples/mplot3d/polys3d.py:7]: That's fair, I changed up the example a bit to take that out. Here's the new image: 
image

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 3]
```diff
@@ -1904,6 +1904,79 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
 
     plot3D = plot
 
+    def fill_between(self, x1, y1, z1, x2, y2, z2, *, where=None, **kwargs):
+        """
+        Fill the area between two 2D or 3D curves.
```
[the remark this answers is not in the document]
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1909]: I was thinking in terms of the example where you're plotting essentially 2D slices and set the 3rd variable to a scalar, but agree saying 2D is probably confusing here. Removed.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 4]
```diff
@@ -1904,6 +1904,79 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+            y1slice = y1[idx0:idx1]
+            z1slice = z1[idx0:idx1]
+            x2slice = x2[idx0:idx1]
+            y2slice = y2[idx0:idx1]
+            z2slice = z2[idx0:idx1]
+
+            if not len(x1slice):
+                continue
+
+            for i in range(len(x1slice) - 1):
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1967]: I think this is in relation to your "mode 2" comment, but I would have naively expected each contiguous region to be a single polygon. Adding additional black edges in the new test case doesn't seem ideal and I would have expected all black "edge colors" to only be around the rim of the fill_between.

I would have expected something like: iterate over all "1" points adding those to the (N*2, 3) list of points for this specific region/poly, move down to "2" and iterate over those in reverse.

I think it will also be much more performant to only have a few polygons instead of one for each quad number of polygons.

**Review by greglucas — CHANGES_REQUESTED** [review]
Requesting changes just so someone doesn't merge without my comment being addressed.

I think "mode 1" vs "mode 2" should be investigated here, and I'd like to see the new test case with the "mode 2" version as well to see if that applies the edge color to only the rim/boundary of the single fill_between region. Naively thinking about it that would be my preference for how to draw these, but I also haven't given it a ton of thought.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 4 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1967]: Ok, I can add a `single_poly` boolean flag (better name welcome) to add back that "mode 2" behavior, as I agree that the single polygon covers a good amount of use cases, looks cleaner with our renderer, and should be more performant. But it's more limiting in the shapes than can be created in the general case, so I would prefer to have it default to `False`.

Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1967]: No difference to the output there reversing the second list of points.

**Comment by scottshambaugh** [comment]
@greglucas I've added the "mode 2" `single_polygon` flag, and the `fillunder3d` example renders nicely as single polygons without artifacts now:
image

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 5]
```diff
@@ -1957,6 +1957,99 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+                x1slice = x1[idx0:idx1]
+                y1slice = y1[idx0:idx1]
+                z1slice = z1[idx0:idx1]
+                x2slice = x2[idx0:idx1]
+                y2slice = y2[idx0:idx1]
+                z2slice = z2[idx0:idx1]
+
+                if not len(x1slice):
+                    continue
+
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/axes3d.py:2039]: I'd put the "single_polygon" case in here, because you still want to break it by the contiguous regions I think? This will simplify the implementation a bit.

```python
            if single_polygon:
                poly = np.stack([np.concatenate([x1slice, x2slice[::-1]]),
                            np.concatenate([y1slice, y2slice[::-1]]),
                            np.concatenate([z1slice, z2slice[::-1]])]).T
                polys.append(poly)
            else:
...
```

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 6]
```diff
@@ -1957,6 +1957,99 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+            Define *where* to exclude some regions from being filled. The
+            filled regions are defined by the coordinates ``pts[where]``,
+            for all x, y, and z pts. More precisely, fill between ``pts[i]``
+            and ``pts[i+1]`` if ``where[i] and where[i+1]``. Note that this
+            definition implies that an isolated *True* value between two
+            *False* values in *where* will not result in filling. Both sides of
+            the *True* position remain unfilled due to the adjacent *False*
+            values.
+
+        single_polygon : bool
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1988]: I think this is a brand new keyword argument and something similar isn't done in any of the other 3D areas? I'm personally not convinced of it and I think it should just be we draw a single polygon and then whatever happens downstream due to our 3d->2d rendering limitations is just that and can be improved later if we get a better renderer. (I also think it should be default True if we do add this as that is what I'd naively expect)

For example, if you do move the image around in your example:
```python
import matplotlib.pyplot as plt
import numpy as np

N = 50
theta = np.linspace(0, 2*np.pi, N)

x1 = np.cos(theta)
y1 = np.sin(theta)
z1 = 0.1 * np.sin(6 * theta)

x2 = 0.6 * np.cos(theta)
y2 = 0.6 * np.sin(theta)
z2 = 2  # Note that scalar values work in addition to length N arrays

fig = plt.figure()
ax = fig.add_subplot(projection='3d')
ax.fill_between(x1, y1, z1, x2, y2, z2, alpha=0.5, edgecolor='k', single_polygon=True)

plt.show()
```
You can see that the `single_polygon` case does look good from the right angle.

**Comment by greglucas** [comment]
Re-requesting @timhoffm's review because it is quite a bit different now with a new keyword argument now and he might have opinions on that from the API standpoint.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 6 above)
Comment (timhoffm) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1988]: It is correct, that this is a new concept, which is inherent to the task. There is no single canonical solution to create a surface between two arbitrary lines. It's thus reasonable to have multiple implementations. However, a bool flag is not good. It it selects or deselects a variant. But does not communicate what is used in the latter case instead. Also there can be other implementations, e.g. triangles instead of quadrangles. Let's use `mode` (or a similar name). I have no clear preference on the default. We could now or later even add a `mode="auto"` that tries to detect the best mode based on the input data.

**Review by timhoffm — DISMISSED** [review]
Please change `single_polygon` to `mode`. I have no clear preference on its default.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 6 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1988]: The single polygon case for that example does not render as intended, it looks like this:
image

From other angles it looks as intended, but this is actually a manifestation of https://github.com/matplotlib/matplotlib/issues/28394 and not correct.
image

Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1988]: I like @timhoffm's suggestion to make this a `mode` selection rather than a flag. Proposed modes are `'auto', 'quad', 'polygon'` (better names suggested). For now I think `'auto'` should map to `'quad'`, since that is a superset of the polygon behavior and gives more flexible results.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 5 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:2039]: Yeah, I think you're right, this matches the 2D `fill_between` behavior better.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 6 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1988]: See also the single polygon mode not performing as intended on the double helix example: 
image

**Comment by scottshambaugh** [comment]
Changed the `single_polygon` flag to a `mode` kwarg, and fixed the `where` behavior for the `'polygon'` mode.

**Comment by scottshambaugh** [comment]
We could try to get smart about `'auto'` selecting `'polygon'` when all the points lie on a plane, but I'm not sure a robust way to do that. A simple case I think would be if one of `(x1 == x2), (y1 == y2), (z1 == z2)` is `True`? Trying to think through potential edge cases there.

**Comment by scottshambaugh** [comment]
The more I think about it the more I like the optimization to map `auto` to `polygon` when all the points are on an x, y, or z plane. It doesn't cover all the cases of points on the same plane, but it will cover what I think will be the most common ones, and the user can always choose the mode directly. Added.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 7]
```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+            If 'auto' and all the coordinates for the x, y, or z input pairs
+            are the same, then the points lie on a plane and 'polygon' is used.
+            Otherwise, 'auto' defaults to 'quad'.
+            If 'quad', then a separate quadrilateral polygon is created for
+            each pair of subsequent points in the 1st and 2nd lines. This is
+            more flexible, but may be slower and result in rendering artifacts.
+            If 'polygon', then the 1st and 2nd lines are connected to form a
+            single polygon, which may be faster and render more cleanly for
+            simple shapes (eg, for filling between two lines that lie on a
+            single plane).
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1999]:
```suggestion
            The fill mode. One of:

            - 'quad':  A separate quadrilateral polygon is created for each
              pair of subsequent points in the two lines.
            - 'polygon': The two lines are connected to fore a single polygon.
              This is faster and can render more cleanly for simple shapes (e.g.
              for filling between two lines that lie within a plane).
            - 'auto': If the lines are in a plane parallel to a coordinate axis
              (one of *x*, *y*. *z* for both lines is a constant), 'polygon' is
              used. Otherwise, 'quad' is used.
```

**Review by timhoffm — COMMENTED** [review]
We may also want to think about `shade`. It's deactivated by default for Poly3dCollection, but activated by default for cases where the drawing is a geometrical object, e.g. in [plot_surface()](https://matplotlib.org/stable/api/_as_gen/mpl_toolkits.mplot3d.axes3d.Axes3D.plot_surface.html) if it's a single color.

Your quad example (at least without edgecolor) could benefit from shading. Without having this thought through completely 'quad' likely wants it on by default, 'polygon' possibly wants it of - there's no shading within the single polygon, but OTOH one could make the polygon example single color and use depth shading to create visual distinction.

Please think about what's the best solution. The default may depend on *mode*. In any case, I think it's valuable to mention it in the docstring.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 6 above)
Comment (greglucas) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: I think I was getting hung up on wanting a known mapping of input values -> polygons so that the collection could be mapped in a known way by the `Collection`/`ScalarMappable`. That is already not the case though because we have the `where` argument and can get any number of `n` polygons being returned based on where the condition holds. So mapping already doesn't really make sense in this case.

I'm satisfied with having both options for constructing the "between" area.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 8]
```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+        `.Poly3DCollection`
+            A `.Poly3DCollection` containing the plotted polygons.
+
+        """
+        _api.check_in_list(['auto', 'quad', 'polygon'], mode=mode)
+
+        had_data = self.has_data()
+        x1, y1, z1, x2, y2, z2 = cbook._broadcast_with_masks(x1, y1, z1, x2, y2, z2)
+        if mode == 'auto':
+            if np.all(x1 == x2) or np.all(y1 == y2) or np.all(z1 == z2):
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/axes3d.py:2015]: > The more I think about it, the more I like the optimization to map auto to polygon when all the points are on an x, y, or z plane. It doesn't cover all the cases of points on an arbitrary plane, but it will cover what I think will be the most common ones, and the user can always choose the mode directly.

How hard would it be to generalize to "if all points lie in a plane"? You could construct a plane from the first 3 points and then calculate a dot product for all following to check distance to that plane and it wouldn't be _too_ bad.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 9]
```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+                mode = 'polygon'
+            else:
+                mode = 'quad'
+
+        if where is None:
+            where = True
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/axes3d.py:2021]: Could this be the default for the keyword? (you might need to broadcast this out in the above call with the points)

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 10]
```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+                            (x1slice[i+1], y1slice[i+1], z1slice[i+1]),
+                            (x2slice[i+1], y2slice[i+1], z2slice[i+1]),
+                            (x2slice[i], y2slice[i], z2slice[i])]
+                    polys.append(poly)
+            elif mode == 'polygon':
+                poly = []
+                for i in range(len(x1slice)):
+                    poly.append((x1slice[i], y1slice[i], z1slice[i]))
+                for i in range(len(x2slice) - 1, -1, -1):
+                    poly.append((x2slice[i], y2slice[i], z2slice[i]))
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/axes3d.py:2053]: We should try to vectorize this if easy enough to do. I think the following will work.
```suggestion
                # Stack the top points in a line
                top = np.column_stack((x1slice, y1slice, z1slice))
                # Stack the second slice in reverse order
                bottom = np.column_stack((x2slice[::-1], y2slice[::-1], z2slice[::-1]))
                # Concatenate the lines into a single poly
                poly = np.concatenate((top, bottom), axis=0)
```

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 11]
```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+            if not len(x1slice):
+                continue
+
+            if mode == 'quad':
+                for i in range(len(x1slice) - 1):
+                    poly = [(x1slice[i], y1slice[i], z1slice[i]),
+                            (x1slice[i+1], y1slice[i+1], z1slice[i+1]),
+                            (x2slice[i+1], y2slice[i+1], z2slice[i+1]),
+                            (x2slice[i], y2slice[i], z2slice[i])]
+                    polys.append(poly)
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/axes3d.py:2047]: Similarly I think this could be vectorized as well, although a little trickier to think about so maybe preallocating is easiest and then filling things in. Something like this:

```python
n_polys = len(x1slice) - 1

# Preallocate the array for storing polygons
polys = np.empty((n_polys, 4, 3))

# First point of quad
polys[:, 0, :] = np.column_stack((x1slice[:-1], y1slice[:-1], z1slice[:-1]))
# Second point
polys[:, 1, :] = np.column_stack((x1slice[1:], y1slice[1:], z1slice[1:]))
# Third point
polys[:, 2, :] = np.column_stack((x2slice[1:], y2slice[1:], z2slice[1:]))
# Fourth point
polys[:, 3, :] = np.column_stack((x2slice[:-1], y2slice[:-1], z2slice[:-1]))
```

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 7 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: > one of *x*, *y*. *z* for both lines is a constant

They need to be constant and equal for this condition to be true, see for a counterexample:

```
n = 50
theta = np.linspace(0, 2*np.pi, n)
x1 = np.cos(theta)
y1 = np.sin(theta)
z1 = 0
x2 = np.cos(theta + np.pi)
y2 = np.sin(theta + np.pi)
z2 = 1
```
image

Updated this docstring to match your format.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 9 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:2021]: The default is `None` in the 2D case and wanted to match that api

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 12]
```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+            for all x, y, and z pts. More precisely, fill between ``pts[i]``
+            and ``pts[i+1]`` if ``where[i] and where[i+1]``. Note that this
+            definition implies that an isolated *True* value between two
+            *False* values in *where* will not result in filling. Both sides of
+            the *True* position remain unfilled due to the adjacent *False*
+            values.
+
+        mode : {'quad', 'polygon', 'auto'}, default: 'auto'
+            The fill mode. One of:
+            - 'quad':  A separate quadrilateral polygon is created for each
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1990]:
```suggestion

            - 'quad':  A separate quadrilateral polygon is created for each
```
Unfortunately, ReST is quite picky and needs an empty line before a bullet list.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 13]
```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
...
+              pair of subsequent points in the two lines.
+            - 'polygon': The two lines are connected to form a single polygon.
+              This is faster and can render more cleanly for simple shapes
+              (e.g. for filling between two lines that lie within a plane).
+            - 'auto': If the lines are in a plane parallel to a coordinate axis
+              (one of *x*, *y*, *z* for both lines are constant and equal),
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1996]: I was confused and going to ask what is "constant but not equal" because constant implies the same value. I was implicitly assuming that constant means the same constant. But if you want that explicitly, I suggest

```suggestion
              (one of *x*, *y*, *z* are constant and equal for both lines),
```

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 8 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: This is working, but is a little clunky... thoughts?

'''python
        i_nonnan = np.where(~np.isnan(x1))[0]
        xs = np.concatenate((x1[i_nonnan], x2[i_nonnan]))
        ys = np.concatenate((y1[i_nonnan], y2[i_nonnan]))
        zs = np.concatenate((z1[i_nonnan], z2[i_nonnan]))

        # calculate the normal vector from the first three points
        n = np.cross([xs[1] - xs[0], ys[1] - ys[0], zs[1] - zs[0]],
                     [xs[2] - xs[0], ys[2] - ys[0], zs[2] - zs[0]])
        # calculate the vectors from the first point to all other points
        vs = np.array([xs - xs[0], ys - ys[0], zs - zs[0]])
        # if the dot product of the normal vector and all other vectors is zero,
        # all points are on the same plane
        dots = np.dot(n, vs)
        all_points_on_plane = np.allclose(dots, 0, atol=1e-12)
'''

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 10 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: I like these vectorizations, done

**Comment by scottshambaugh** [comment]
@timhoffm I do like the addition of the `shade` parameter, and think it should be on by default for quad and off by default for polygon. AFAIK depthshading is for `Patch3DCollection` we don't have it built in for `Poly3DCollection`s?

Here's the top example with shading and no transparency:
image

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 8 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: @greglucas I added `art3d._all_points_on_plane()` to check for this, it's a little involved to catch the edge cases - what do you think? I still need to add tests but wanted to get a pulse first.

File: lib/mpl_toolkits/mplot3d/tests/test_art3d.py [hunk 14]
```diff
@@ -54,3 +54,25 @@ def test_zordered_error():
     ax.add_collection(Line3DCollection(lc))
     ax.scatter(*pc, visible=False)
     plt.draw()
+
+
+def test_all_points_on_plane():
```
Comment (greglucas) [inline lib/mpl_toolkits/mplot3d/tests/test_art3d.py:59]: Probably OK as is here, but you could parametrize these cases as well.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 8 above)
Comment (greglucas) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: Sorry for sending you down the rabbit hole 😂 . I think it is great! and I really appreciate that you checked the edge cases. Maybe it is a little involved, but seems worth it to me to generalize the statement to any plane. It is only a one-time performance hit when creating the collection initially and it is all numpy-vectorized so it isn't that terrible.

**Review by greglucas — APPROVED** [review]
I think this looks good now and there is nice auto-handling of all the various cases.

**Comment by scottshambaugh** [comment]
CI failures are unrelated

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Thanks @scottshambaugh. Great addition!
